src/aws/flb_aws_compress.c

```diff
@@ -0,0 +1,231 @@
+ */
+
+#include <fluent-bit/flb_mem.h>
+#include <fluent-bit/flb_log.h>
+#include <mbedtls/base64.h>
+
+#include <fluent-bit/aws/flb_aws_compress.h>
+#include <fluent-bit/flb_gzip.h>
+
+#include <stdint.h>
+
+#ifdef FLB_HAVE_ARROW
+#include "compression/arrow/compress.h"
+#endif
+
+struct compression_option {
+    int compression_type;
+    char *compression_keyword;
+    int(*compress)(void *in_data, size_t in_len, void **out_data, size_t *out_len);
+};
+
+/*
+ * Library of compression options
+ * AWS plugins that support compression will have these options.
+ * Referenced function should return -1 on error and 0 on success.
+ */
+static const struct compression_option compression_options[] = {
+    /* FLB_AWS_COMPRESS_NONE which is 0 is reserved for array footer */
+    {
+        FLB_AWS_COMPRESS_GZIP,
+        "gzip",
+        &flb_gzip_compress
+    },
+#ifdef FLB_HAVE_ARROW
+    {
+        FLB_AWS_COMPRESS_ARROW,
+        "arrow",
+        &out_s3_compress_arrow
+    },
+#endif
+    { 0 }
+};
+
+int flb_aws_compression_get_type(const char *compression_keyword)
+{
+    int ret;
+    const struct compression_option *o;
+
+    o = compression_options;
+
+    while (o->compression_type != 0) {
+        ret = strcmp(o->compression_keyword, compression_keyword);
+        if (ret == 0) {
+            return o->compression_type;
+        }
+        ++o;
+    }
+
+    flb_error("[aws_compress] unknown compression type: %s", compression_keyword);
+    return -1;
+}
+
+int flb_aws_compression_compress(int compression_type, void *in_data, size_t in_len,
+                                void **out_data, size_t *out_len)
+{
+    const struct compression_option *o;
+
+    o = compression_options;
+
+    while (o->compression_type != 0) {
+        if (o->compression_type == compression_type) {
+            return o->compress(in_data, in_len, out_data, out_len);
+        }
+        ++o;
+    }
+    
+    flb_error("[aws_compress] invalid compression type: %i", compression_type);
+    flb_errno();
+    return -1;
+}
+
+int flb_aws_compression_b64_truncate_compress(int compression_type, size_t max_out_len,
+                                             void *in_data, size_t in_len,
+                                             void **out_data, size_t *out_len)
+{
+    static const void *truncation_suffix = "[Truncated...]";
+    static const size_t truncation_suffix_len = 14;
+    static const double truncation_reduction_percent = 90; /* % out of 100 */
+
+    int ret;
+    int is_truncated;
+    size_t truncated_in_len_prev;
+    size_t truncated_in_len;
+    void *truncated_in_buf;
+    void *compressed_buf;
+    size_t compressed_len;
+    size_t original_b64_compressed_len;
+
+    unsigned char *b64_compressed_buf;
+    size_t b64_compressed_len;
+    size_t b64_actual_len;
+
+    /* Iterative approach to truncation */
+    truncated_in_len = in_len;
+    truncated_in_buf = in_data;
+    is_truncated = FLB_FALSE;
+    b64_compressed_len = SIZE_MAX;
```

Comment:
where does SIZE_MAZ come from??

Reply:
SIZE_MAX comes from stdint.h providing a Limit of size_t type
```
/* Limit of `size_t' type.  */
# if __WORDSIZE == 64
#  define SIZE_MAX		(18446744073709551615UL)
# else
#  if __WORDSIZE32_SIZE_ULONG
#   define SIZE_MAX		(4294967295UL)
#  else
#   define SIZE_MAX		(4294967295U)
#  endif
# endif
```

Please see: https://en.cppreference.com/w/c/types/size_t

Reply:
I can do  ( (size_t) -1 ) if SIZE_MAX is not standard enough